haas/api.py

```diff
@@ -258,34 +275,57 @@ def node_connect_network(node, nic, network):
 
     project = node.project
 
+    allocator = get_network_allocator()
+    if channel is None:
+        channel = allocator.get_default_channel(db)
+
     if nic.current_action:
         raise BlockedError("A networking operation is already active on the nic.")
 
     if (network.access is not None) and (network.access is not project):
         raise ProjectMismatchError("Project does not have access to given network.")
 
-    db.add(model.NetworkingAction(nic, network))
+    if _have_attachment(db, nic, model.NetworkAttachment.network == network):
+        raise BlockedError("The network is already attached to the nic.")
+    if _have_attachment(db, nic, model.NetworkAttachment.channel == channel):
+        raise BlockedError("The channel is already in use on the nic.")
+
+    if not allocator.is_legal_channel_for(db, channel, network.network_id):
+        raise BadArgumentError("Channel %r, is not legal for this network." %
+                               channel)
+
+    db.add(model.NetworkingAction(nic=nic,
+                                  new_network=network,
+                                  channel=channel))
     db.commit()
 
 @rest_call('POST', '/node/<node>/nic/<nic>/detach_network')
-def node_detach_network(node, nic):
-    """Detach a physical nic from any network it's on.
+def node_detach_network(node, nic, network):
+    """Detach network ``network`` from physical nic ``nic``.
 
     Raises ProjectMismatchError if the node is not in a project.
+
+    Raises BlockedError if there is already a pending network action.
+
+    Raises BadArgumentError if the network is not attached to the nic.
     """
     db = model.Session()
     node = _must_find(db, model.Node, node)
+    network = _must_find(db, model.Network, network)
     nic = _must_find_n(db, node, model.Nic, nic)
 
     if not node.project:
         raise ProjectMismatchError("Node not in project")
 
-    project = nic.owner.project
-
     if nic.current_action:
         raise BlockedError("A networking operation is already active on the nic.")
-
-    db.add(model.NetworkingAction(nic, None))
+    attachment = db.query(model.NetworkAttachment)\
+        .filter_by(nic=nic, network=network).first()
+    if attachment is None:
+        raise BadArgumentError("The network is not attached to the nic.")
+    db.add(model.NetworkingAction(nic=nic,
+                                  channel=attachment.channel,
+                                  new_network=None))
     db.commit()
```

Comment:
The (new) documentation says node_connect/detach_network returns 202 Accepted. Does this do that?